File: packages/web-game-client/src/game/engine.ts
// packages/web-game-client/src/game/engine.ts

import { GameState, PlayerState, Action, Card, CardTemplate, ResolvedAction, EffectAction, Deck } from '../types';
import { acquireProperty } from './effects/acquireProperty';
import { gainFunds } from './effects/gainFunds';

// A helper type for the resolution process
interface EffectContext {
  player: PlayerState;
  opponent: PlayerState;
  card: CardTemplate;
  isCancelled: boolean;
}

export class GameEngine {
  private state: GameState;
  private cardTemplates: { [key: string]: CardTemplate };

  constructor(initialState: GameState, cardTemplates: { [key: string]: CardTemplate }) {
    if (!cardTemplates || Object.keys(cardTemplates).length === 0) {
      console.error("GameEngine: cardTemplates is empty or undefined.", cardTemplates);
      throw new Error("GameEngine must be initialized with valid cardTemplates.");
    }
    // State Hydration: Ensure arrays exist
    initialState.players.forEach(player => {
      player.hand = player.hand || [];
      player.deck = player.deck || [];
      player.discard = player.discard || [];
    });
    initialState.lastActions = initialState.lastActions || [];
    this.state = JSON.parse(JSON.stringify(initialState));
    this.cardTemplates = cardTemplates;
  }

  public getState(): GameState {
    return JSON.parse(JSON.stringify(this.state));
  }

  public setState(newState: GameState): void {
    this.state = JSON.parse(JSON.stringify(newState));
  }

  public applyAction(player1Action: Action | null, player2Action: Action | null): GameState {
    if (this.state.phase === 'GAME_OVER') return this.getState();
    const newState = this.resolveActions(this.getState(), player1Action, player2Action);
    this.checkWinCondition(newState); // Check win condition AFTER actions are resolved
    this.state = newState;
    return this.getState();
  }

  public advanceTurn(): GameState {
    if (this.state.phase === 'GAME_OVER') return this.getState();

    this.state.turn++;
    this.state.phase = 'DRAW';
    const playerFunds = this.state.players[0].funds;
    const opponentFunds = this.state.players[1].funds;
    this.state.log.push(`ターン${this.state.turn}　プレイヤー資産：${playerFunds} 対戦相手資産：${opponentFunds}`);

    this.state.players.forEach(player => {
      player.discard.push(...player.hand);
      player.hand = [];
    });

    this.state.players.forEach(player => {
      this.drawCards(player, 3);
    });

    this.state.phase = 'ACTION';
    this.state.lastActions = [];
    return this.getState();
  }

  public getCardTemplate(templateId: string): CardTemplate | undefined {
    return this.cardTemplates[templateId];
  }

  public getPlayableCards(player: PlayerState): Card[] {
    if (!player) return [];
    return player.hand.filter(card => {
      const template = this.getCardTemplate(card.templateId);
      return template && player.funds >= template.cost;
    });
  }

  public static createInitialState(player1Id: string, player2Id: string, cardTemplates: { [key: string]: CardTemplate }, player1Deck: Deck, player2Deck: Deck): GameState {
    const createPlayer = (id: string, deckData: Deck): PlayerState => {
      const cardsInDeck: Card[] = [];
      for (const templateId in deckData.cards) {
        for (let i = 0; i < deckData.cards[templateId]; i++) {
          cardsInDeck.push({ uuid: crypto.randomUUID(), templateId });
        }
      }
      return {
        playerId: id, funds: 0, properties: 1, hand: [],
        deck: cardsInDeck.sort(() => Math.random() - 0.5),
        discard: [],
      };
    };
    return {
      matchId: `match-${Date.now()}`,
      turn: 0,
      players: [createPlayer(player1Id, player1Deck), createPlayer(player2Id, player2Deck)],
      phase: 'DRAW',
      lastActions: [],
      result: 'IN_PROGRESS',
      log: ['ゲーム開始！'],
    };
  }

  private drawCards(player: PlayerState, count: number): void {
    for (let i = 0; i < count; i++) {
      if (player.deck.length === 0) {
        if (player.discard.length === 0) break;
        player.deck = [...player.discard].sort(() => Math.random() - 0.5);
        player.discard = [];
      }
      const drawnCard = player.deck.shift();
      if (drawnCard) player.hand.push(drawnCard);
    }
  }

  private resolveActions(state: GameState, p1Action: Action | null, p2Action: Action | null): GameState {
    let mutableState: GameState = JSON.parse(JSON.stringify(state));
    mutableState.phase = 'RESOLUTION';

    const p1 = mutableState.players.find(p => p.playerId === p1Action?.playerId);
    const p2 = mutableState.players.find(p => p.playerId === p2Action?.playerId);

    const getEffectContext = (player: PlayerState | undefined, action: Action | null, opponent: PlayerState | undefined): EffectContext | null => {
      if (!action || !player) return null;

      let cardInstance: Card | undefined;
      let template: CardTemplate | undefined;

      if (action.actionType === 'collect_funds') {
        template = this.cardTemplates['COLLECT_FUNDS'];
        if (!template) return null;
      } else {
        cardInstance = player.hand.find(c => c.uuid === action.cardUuid);
        if (!cardInstance) return null;
        template = this.cardTemplates[cardInstance.templateId];
        if (!template) return null;
      }

      if (player.funds < template.cost) return null;
      
      return { player, opponent: opponent || player, card: template, isCancelled: false };
    };

    const p1Context = getEffectContext(p1, p1Action, p2);
    const p2Context = getEffectContext(p2, p2Action, p1);

    mutableState.lastActions = [p1Context, p2Context]
      .filter((context): context is EffectContext => !!context)
      .map(context => ({ playerId: context.player.playerId, cardTemplateId: context.card.templateId }));

    [p1Context, p2Context].forEach(context => {
      if (context && context.card.templateId !== 'COLLECT_FUNDS') {
        context.player.funds -= context.card.cost;
        const cardInstance = context.player.hand.find(c => c.templateId === context.card.templateId);
        if (cardInstance) {
          context.player.hand = context.player.hand.filter(c => c.uuid !== cardInstance.uuid);
          context.player.discard.push(cardInstance);
        }
      }
    });

    if (p1Context && p2Context) {
        const p1Card = p1Context.card.templateId;
        const p2Card = p2Context.card.templateId;

        // --- Fraud-specific Interactions first ---
        if (p1Card === 'FRAUD') {
            if (p2Card === 'ACQUIRE') { p2Context.isCancelled = true; } else { p1Context.isCancelled = true; }
        }
        if (p2Card === 'FRAUD') {
            if (p1Card === 'ACQUIRE') { p1Context.isCancelled = true; } else { p2Context.isCancelled = true; }
        }

        // --- Standard Interactions ---
        if (p1Card === 'DEFEND' && p2Card === 'ACQUIRE') { p2Context.isCancelled = true; }
        if (p2Card === 'DEFEND' && p1Card === 'ACQUIRE') { p1Context.isCancelled = true; }
        if (p1Card === 'BRIBE' && p2Context.card.effect.category === 'DEFENSE') { p2Context.isCancelled = true; }
        if (p2Card === 'BRIBE' && p1Context.card.effect.category === 'DEFENSE') { p1Context.isCancelled = true; }
        
        // --- Mutual Cancellation ---
        if (p1Card === p2Card && (p1Card === 'ACQUIRE' || p1Card === 'BRIBE')) {
            p1Context.isCancelled = true;
            p2Context.isCancelled = true;
            mutableState.log.push('両者の行動は互いに打ち消しあった！');
        }
    }

    const contextsToApply = [p1Context, p2Context]
        .filter((c): c is EffectContext => c !== null && !c.isCancelled)
        .sort((a, b) => b.card.effect.priority - a.card.effect.priority);

    for (const context of contextsToApply) {
      for (const effectAction of context.card.effect.actions) {
        mutableState = this.applyEffect(mutableState, context.player.playerId, context.opponent.playerId, effectAction);
      }
      const actionName = context.card.templateId === 'COLLECT_FUNDS' ? '資金集め' : context.card.name;
      mutableState.log.push(`${context.player.playerId === mutableState.players[0].playerId ? 'プレイヤー' : '対戦相手'}の行動「${actionName}」`);
    }

    return mutableState;
  }

  private checkConditions(state: GameState, player: PlayerState, opponent: PlayerState, card: CardTemplate, allPlayedActions: ResolvedAction[]): boolean {
    const conditions = card.effect.conditions;
    if (!conditions) return true;

    if (conditions.hasFunds && player.funds < conditions.hasFunds) return false;
    if (conditions.opponentPlayedCard) {
      const opponentAction = allPlayedActions.find(action => action.playerId === opponent.playerId);
      if (opponentAction?.cardTemplateId !== conditions.opponentPlayedCard) return false;
    }
    if (conditions.opponentHasProperty && opponent.properties < conditions.opponentHasProperty) return false;

    return true;
  }

  private applyEffect(state: GameState, actingPlayerId: string, opponentPlayerId: string, action: EffectAction): GameState {
    const player = state.players.find(p => p.playerId === actingPlayerId);
    if (!player) return state;

    switch (action.name) {
      case 'ACQUIRE_PROPERTY':
        return acquireProperty(state, actingPlayerId, opponentPlayerId, action);
      case 'GAIN_FUNDS':
        return gainFunds(state, actingPlayerId, opponentPlayerId, action);
      case 'CANCEL_EFFECT':
        // Handled in interaction logic
        break;
    }
    return state;
  }

  private checkWinCondition(state: GameState): void {
    const p1Lost = state.players[0].properties <= 0;
    const p2Lost = state.players[1].properties <= 0;

    if (p1Lost || p2Lost) {
      state.phase = 'GAME_OVER'; // Set phase to GAME_OVER
      if (p1Lost && p2Lost) {
        state.result = 'DRAW';
        state.log.push('両者、同時に不動産をすべて失った！引き分け！');
      } else if (p1Lost) {
        state.result = 'LOSE';
        state.log.push(`プレイヤーの不動産が0になった！対戦相手の勝利！`);
      } else { // p2Lost
        state.result = 'WIN';
        state.log.push(`対戦相手の不動産が0になった！プレイヤーの勝利！`);
      }
    }
  }
}

File: packages/api-server/app/game/engine.py
import random
import time
from typing import Dict, List, Optional

from app.game.cards import (apply_acquire, apply_defend, apply_fraud,
                            apply_gain_funds)
from app.game.models import (Action, Card, CardTemplate, GameState,
                             PlayerState, ResolvedAction)


class GameEngine:
    def __init__(self, initial_state: GameState, card_templates: Dict[str, CardTemplate]):
        # Python Pydantic models handle initialization, so no need for hydration
        self.state = initial_state
        self.card_templates = card_templates

    def get_state(self) -> GameState:
        return self.state.copy(deep=True)

    def apply_action(self, player1_action: Optional[Action], player2_action: Optional[Action]) -> GameState:
        if self.state.phase == 'GAME_OVER':
            return self.get_state()

        new_state = self.get_state()
        
        resolved_actions = self._resolve_actions(new_state, player1_action, player2_action)
        new_state.lastActions = resolved_actions
        
        self._check_win_condition(new_state)
        
        self.state = new_state
        return self.get_state()

    def advance_turn(self) -> GameState:
        if self.state.phase == 'GAME_OVER':
            return self.get_state()

        self.state.turn += 1
        self.state.phase = 'DRAW'
        self.state.lastActions = []
        self.state.log.append(f"--- ターン {self.state.turn} ---")

        for player in self.state.players:
            cards_to_draw = 3 - len(player.hand)
            if cards_to_draw > 0:
                self._draw_cards(player, cards_to_draw)
        
        self.state.phase = 'ACTION'
        return self.get_state()

    def get_card_template(self, template_id: str) -> Optional[CardTemplate]:
        return self.card_templates.get(template_id)

    @staticmethod
    def create_initial_state(player1_id: str, player2_id: str, card_templates: Dict[str, CardTemplate]) -> GameState:
        initial_deck = []
        for t in card_templates.values():
            count = 4 if t.name == '資金集め' else 2
            for _ in range(count):
                initial_deck.append(t.templateId)

        def create_player(p_id: str) -> PlayerState:
            deck_cards = [Card(id=f"card{i}_{tid}", templateId=tid) for i, tid in enumerate(initial_deck)]
            random.shuffle(deck_cards)
            return PlayerState(
                playerId=p_id,
                funds=2,
                properties=1,
                deck=deck_cards
            )

        return GameState(
            matchId=f"match-{int(time.time())}",
            turn=0,
            players=[create_player(player1_id), create_player(player2_id)],
            phase='DRAW',
            log=['ゲーム開始！']
        )

    def _draw_cards(self, player: PlayerState, count: int) -> None:
        for _ in range(count):
            if not player.deck:
                if not player.discard:
                    break
                player.deck = player.discard
                player.discard = []
                random.shuffle(player.deck)
            
            drawn_card = player.deck.pop(0)
            player.hand.append(drawn_card)

    def _apply_card_effect(self, state: GameState, player: PlayerState, card: Card, opponent: PlayerState) -> None:
        card_template = self.get_card_template(card.templateId)
        if not card_template:
            return

        if card_template.type == 'GAIN_FUNDS':
            apply_gain_funds(player)
        elif card_template.type == 'ACQUIRE':
            apply_acquire(player, opponent)
        elif card_template.type == 'DEFEND':
            apply_defend(player)
        elif card_template.type == 'FRAUD':
            apply_fraud(player, opponent)

    def _resolve_actions(self, state: GameState, player1_action: Optional[Action], player2_action: Optional[Action]) -> List[ResolvedAction]:
        state.phase = 'RESOLUTION'
        resolved = []

        player1 = next((p for p in state.players if p.playerId == player1_action.playerId), None) if player1_action else None
        player2 = next((p for p in state.players if p.playerId == player2_action.playerId), None) if player2_action else None

        def get_card_info(player: Optional[PlayerState], action: Optional[Action]):
            if not player or not action:
                return None, None
            card = next((c for c in player.hand if c.id == action.cardId), None)
            template = self.get_card_template(card.templateId) if card else None
            return card, template

        p1_card, p1_template = get_card_info(player1, player1_action)
        p2_card, p2_template = get_card_info(player2, player2_action)

        p1_played = p1_card and p1_template and player1 and player1.funds >= p1_template.cost
        p2_played = p2_card and p2_template and player2 and player2.funds >= p2_template.cost

        if p1_played:
            player1.funds -= p1_template.cost
            player1.hand = [c for c in player1.hand if c.id != p1_card.id]
            player1.discard.append(p1_card)
            resolved.append(ResolvedAction(playerId=player1.playerId, cardTemplateId=p1_template.templateId))
            state.log.append(f"プレイヤーは「{p1_template.name}」をプレイした")

        if p2_played:
            player2.funds -= p2_template.cost
            player2.hand = [c for c in player2.hand if c.id != p2_card.id]
            player2.discard.append(p2_card)
            resolved.append(ResolvedAction(playerId=player2.playerId, cardTemplateId=p2_template.templateId))
            state.log.append(f"対戦相手は「{p2_template.name}」をプレイした")

        p1_effective_template = p1_template if p1_played else None
        p2_effective_template = p2_template if p2_played else None

        is_p1_acquire = p1_effective_template and p1_effective_template.type == 'ACQUIRE'
        is_p2_acquire = p2_effective_template and p2_effective_template.type == 'ACQUIRE'
        is_p1_defend = p1_effective_template and p1_effective_template.type == 'DEFEND'
        is_p2_defend = p2_effective_template and p2_effective_template.type == 'DEFEND'
        is_p1_fraud = p1_effective_template and p1_effective_template.type == 'FRAUD'
        is_p2_fraud = p2_effective_template and p2_effective_template.type == 'FRAUD'

        p1_effect = bool(p1_played)
        p2_effect = bool(p2_played)

        if is_p1_acquire and is_p2_acquire:
            p1_effect = p2_effect = False
        elif is_p1_acquire and is_p2_defend:
            p1_effect = False
        elif is_p1_acquire and is_p2_fraud:
            p1_effect = False
            opponent_of_p2 = next(p for p in state.players if p.playerId != player2.playerId)
            self._apply_card_effect(state, player2, p2_card, opponent_of_p2)
        elif is_p2_acquire and is_p1_defend:
            p2_effect = False
        elif is_p2_acquire and is_p1_fraud:
            p2_effect = False
            opponent_of_p1 = next(p for p in state.players if p.playerId != player1.playerId)
            self._apply_card_effect(state, player1, p1_card, opponent_of_p1)

        if p1_effect and p1_effective_template and p1_effective_template.type not in ['FRAUD', 'DEFEND']:
            opponent_of_p1 = next(p for p in state.players if p.playerId != player1.playerId)
            self._apply_card_effect(state, player1, p1_card, opponent_of_p1)
        
        if p2_effect and p2_effective_template and p2_effective_template.type not in ['FRAUD', 'DEFEND']:
            opponent_of_p2 = next(p for p in state.players if p.playerId != player2.playerId)
            self._apply_card_effect(state, player2, p2_card, opponent_of_p2)

        return resolved

    def _check_win_condition(self, state: GameState) -> None:
        p1_lost = state.players[0].properties <= 0
        p2_lost = state.players[1].properties <= 0

        if p1_lost or p2_lost:
            state.phase = 'GAME_OVER'
            print('ゲーム終了')